src/accelerate/hooks.py

```diff
@@ -155,17 +155,17 @@ def add_hook_to_module(module: nn.Module, hook: ModelHook, append: bool = False)
     module = hook.init_hook(module)
     module._hf_hook = hook
 
-    @functools.wraps(old_forward)
-    def new_forward(*args, **kwargs):
+    def new_forward(module, *args, **kwargs):
         args, kwargs = module._hf_hook.pre_forward(module, *args, **kwargs)
         if module._hf_hook.no_grad:
             with torch.no_grad():
-                output = old_forward(*args, **kwargs)
+                output = module._old_forward(*args, **kwargs)
         else:
-            output = old_forward(*args, **kwargs)
+            output = module._old_forward(*args, **kwargs)
         return module._hf_hook.post_forward(module, output)
 
-    module.forward = new_forward
+    module.forward = functools.update_wrapper(functools.partial(new_forward, module), old_forward)
```

Comment:
I wonder about this line versus:

```python
module.forward = MethodType(functools.update_wrapper(new_forward, old_forward), module)
```

@austinapatel Do you think that would be preferable? I thought it might affect pickling, but unfortunately, both variants don't allow it, but the error messages are different.

Reply:
Thanks for the suggestion and review! I just tried your code snippet, and I run into the same issue as this comment: https://github.com/huggingface/accelerate/pull/1971#discussion_r1326765316 since `functools.update_wrapper` needs to be called last (`test_add_and_remove_hooks` in `test_hooks.py` fails). If I try `module.forward = functools.update_wrapper(MethodType(new_forward, module), old_forward)` I get the following error: `AttributeError: 'method' object has no attribute '__module__'` when calling `update_wrapper`. My understanding is that `update_wrapper` is now being called on a method rather than on a function, which doesn't play as nicely with `update_wrapper`.

Reply:
Thanks for explaining and testing, this makes sense. I'm a bit wary about binding the module via `partial` to basically imitate a bound method, but it seems that all other solutions don't work, so I guess it's the best we can do.